pac4j-cas/src/main/java/org/pac4j/cas/client/rest/AbstractCasRestClient.java

```diff
@@ -0,0 +1,120 @@
+import org.pac4j.core.client.BaseClient;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.exception.TechnicalException;
+import org.pac4j.http.client.direct.DirectHttpClient;
+import org.pac4j.http.credentials.UsernamePasswordCredentials;
+import org.pac4j.http.profile.HttpProfile;
+import org.pac4j.http.profile.creator.AuthenticatorProfileCreator;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+/**
+ * This is {@link CasRestFormClient} able to communicate to
+ * a CAS server via its REST api, to retrieve TGTs and STs
+ * and to construct CAS principals.
+ *
+ * @author Misagh Moayyed
+ * @since 1.8.0
+ */
+public abstract class AbstractCasRestClient extends DirectHttpClient<UsernamePasswordCredentials> {
+    public AbstractCasRestClient() {
+        super();
+    }
+
+    public AbstractCasRestClient(final CasRestAuthenticator authenticator) {
+        setAuthenticator(authenticator);
+        setProfileCreator(new AuthenticatorProfileCreator<UsernamePasswordCredentials, HttpProfile>());
+    }
+
+    public HttpTGTProfile requestTicketGrantingTicket(final WebContext context) {
+        final UsernamePasswordCredentials creds = this.extractor.extract(context);
+        getAuthenticator().validate(creds);
+        return (HttpTGTProfile) getProfileCreator().create(creds);
+    }
+
+    public CasCredentials requestServiceTicket(final String serviceURL, final HttpTGTProfile profile) {
+        HttpURLConnection connection = null;
+        try {
+            final URL endpointURL = new URL(getAuthenticator().getCasRestUrl());
+            final URL ticketURL = new URL(endpointURL, endpointURL.getPath() + "/" + profile.getTicketGrantingTicketId());
+
+            connection = HttpUtils.openConnection(ticketURL);
+            final String payload = HttpUtils.encodeQueryParam("service", serviceURL);
+
+            final BufferedWriter out = new BufferedWriter(new OutputStreamWriter(connection.getOutputStream()));
+            out.write(payload);
+            out.close();
+
+            final int responseCode = connection.getResponseCode();
+            if (responseCode == HttpStatus.SC_OK) {
+                final BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()));
+                return new CasCredentials(in.readLine(), getClass().getSimpleName());
+            }
+            throw new TechnicalException("Service ticket request for `" + profile + "` failed: " +
+                    HttpUtils.buildHttpErrorMessage(connection));
+        } catch (final IOException e) {
+            throw new TechnicalException(e);
+        } finally {
+            HttpUtils.closeConnection(connection);
+        }
+    }
+
+    public CasProfile validateServiceTicket(final String serviceURL, final CasCredentials ticket) {
```

Comment:
Do we really need this one? Will we really validate internally the service ticket we create? I don't think so. I guess we will call web services using service tickets, won't we?